Matchmaker integration. Complete

northwestwitch opened the pull request:
This PR contains the entire code that adds/updates or removes patients from patientMatcher, sends internal match requests to match scout patients or requests to external nodes. It has also an interface to visualize the submitted data  internal and external matches results.

Instructions to test this PR:

1. Install patientMatcher(https://github.com/Clinical-Genomics/patientMatcher). There's no need to create username and password for its database for testing purposes, so skip Prerequisites.

2. Using the patientMatcher cli, add a client for scout as described here: https://github.com/Clinical-Genomics/patientMatcher#adding-a-client-to-the-database. For instance you could use this command:
```bash
pmatcher add client -id scout_test -token test_token -url scout_test_url
```

3. Run patientMatcher on a different port than the one you use to run scout. For instance:
```bash
pmatcher run -p 9020
```

4. Make sure that you don't see anything matchmaker-related in the case page of scout.

5. Add a 'mme_submitter' role to your scout test user:
![image](https://user-images.githubusercontent.com/28093618/52401714-aa423b00-2ac2-11e9-9313-473dd09e336d.png)

6. Edit the scout config file(in server/config.py) adding these params:
#MatchMaker related parameters
MME_ACCEPTS = 'application/vnd.ga4gh.matchmaker.v1.0+json'
MME_URL = 'http://localhost:9020'
MME_TOKEN = 'test_token' #this should be the same of when you created the client in step 2

7. Go to the case page and see that now you can submit max 3 pinned variants to matchmaker (genes only or complete variants). You can optionally submit patient sex, hpo terms and OMIM diagnoses. The form should look like this:
![image](https://user-images.githubusercontent.com/28093618/52900867-05c7a380-31fc-11e9-849a-455caf32634b.png)

8. Once you submit the patient you can update your submission by changing pinned variants, or the other options described above. You should also be able to delete the patient from matchmaker. The form should look like this:
![image](https://user-images.githubusercontent.com/28093618/52900878-39a2c900-31fc-11e9-8b14-a9c96e5a71b4.png)

9. From the same interface you will be able to run matches against scout patients in MatchMaker:
![image](https://user-images.githubusercontent.com/28093618/52901008-c306cb00-31fd-11e9-83a4-329412165840.png)
When testing on stage we will be connected to an external node and the option will be showed here.

10. Once having runned one of more match request against the internal node check that there are results in the matches interface:
![image](https://user-images.githubusercontent.com/28093618/52901044-3d374f80-31fe-11e9-9e02-26143865ddb2.png)

northwestwitch commented:
I'm done with this. Whoever reviews gets a present!  :)

dnil commented:
Yay! This looks like really good stuff. I will see what I can do in the afternoon, if noone beats me to it.

northwestwitch commented:
I don't think there is a line of people interested in reviewing this monster.. Thanks! :)

moonso commented:
Wow! Nice work @northwestwitch 🤸‍♀️ . I will work on CG a couple of days this week so will have a look then

northwestwitch commented:
Nice, thanks!

northwestwitch commented:
I'm doing some fixes to this PR because I've changed some stuff in patientMatcher last week, I'll let you know when this is ready!

northwestwitch commented:
I think I'm done with this!

dnil commented:
Ok, so, finally here. Having some issues getting my local setup to run though:
- pmatcher is running, up to date with master. Running ok as far as I can tell: 
```
$ pmatcher --version
Flask 1.0.2
Python 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
$ pmatcher run -h localhost -p 9020
```
Scout connects, but is grumpy about the response it receives: 
```
 Sending GET request to MME url http://localhost:9020/nodes. Data sent: {'timestamp': 1553852611.830226}
2019-03-29 10:43:31 k1315C02VR0WRHTDH.local urllib3.connectionpool[11344] DEBUG Starting new HTTP connection (1): localhost:9020
INFO:patientMatcher.auth.auth:Authorized client with id "scout_test" submits a GET request
INFO:patientMatcher.server.views:Authorized client requests list of connected nodes..
INFO:werkzeug:127.0.0.1 - - [29/Mar/2019 10:43:31] "GET /nodes HTTP/1.1" 200 -
2019-03-29 10:43:31 k1315C02VR0WRHTDH.local urllib3.connectionpool[11344] DEBUG http://localhost:9020 "GET /nodes HTTP/1.1" 200 3
2019-03-29 10:43:31 k1315C02VR0WRHTDH.local scout.utils.matchmaker[11344] INFO MME server response was:[]
2019-03-29 10:43:31 k1315C02VR0WRHTDH.local scout.utils.matchmaker[11344] INFO An error occurred while sending HTTP request to server (list indices must be integers or slices, not str)
```
I've tried disabling pmatcher: no connection, and Scout gives it up after a few retries, so I assume connection is more or less ok. Looks like there are no nodes - confirmed: 
```
curl -X GET   -H 'X-Auth-Token: test_token'   localhost:9020/nodes INFO:patientMatcher.auth.auth:Authorized client with id "scout_test" submits a GET request
INFO:patientMatcher.server.views:Authorized client requests list of connected nodes.. INFO:werkzeug:127.0.0.1 - - [29/Mar/2019 10:55:16] "GET /nodes HTTP/1.1" 200 -
[]
```
Is this expected? If so, this should not return an error. Should I add a fake/local node to get it working?

dnil commented on scout/parse/matchmaker.py:
```diff
@@ -0,0 +1,226 @@
+            zygosities = var['samples'] # it's a list with zygosity situation for each sample of the case
+            for zyg in zygosities:
+                if zyg.get('display_name') == sample_name: # sample of interest
+                    zygosity = zyg['genotype_call'].count('1')
```
Couldn't it be 1/2?

northwestwitch replied:
I guess so, I'll fix it!

dnil commented on scout/utils/matchmaker.py:
```diff
@@ -0,0 +1,76 @@
+        return nodes
+    url = ''.join([mme_base_url, '/nodes'])
+    nodes = matchmaker_request(url=url, token=token, method='GET')
+    if isinstance(nodes,dict):
```
Maybe a comment to say why this check?

northwestwitch replied:
I've realized that whatever was passed to the function was always a dict, and it was wrong. I've fixed now!

northwestwitch commented:
Wow, you started on this, nice!!!

dnil commented:
Trying it after manually removing the case from the pmatcher db, this is the error that occurs around adding time:
```
019-03-29 12:12:16 k1315C02VR0WRHTDH.local scout.adapter.mongo.hgnc[11496] DEBUG Fetching gene 17284
2019-03-29 12:12:16 k1315C02VR0WRHTDH.local scout.utils.matchmaker[11496] INFO Sending POST request to MME url http://localhost:9020/patient/add. Data sent: {'patient': {'contact': {'name': 'Clark Kent', 'href': 'mailto:clark.kent@mail.com', 'institution': 'Scout software user, Science For Life Laboratory, Stockholm, Sweden'}, 'id': '643594.ADM1059A2', 'label': '643594.NA12882', 'features': [{'id': 'HP:0002664', 'label': 'Neoplasm', 'observed': 'yes'}, {'id': 'HP:0000014', 'label': 'Abnormality of the bladder', 'observed': 'yes'}, {'id': 'HP:0001222', 'label': 'Spatulate thumbs', 'observed': 'yes'}, {'id': 'HP:0009733', 'label': 'Glioma', 'observed': 'yes'}, {'id': 'HP:0000510', 'label': 'Rod-cone dystrophy', 'observed': 'yes'}, {'id': 'HP:0001251', 'label': 'Ataxia', 'observed': 'yes'}, {'id': 'HP:0000118', 'label': 'Phenotypic abnormality', 'observed': 'yes'}], 'disorders': [], 'sex': 'MALE', 'genomicFeatures': [{'gene': {'id': 'POT1'}, 'variant': {'referenceName': '7', 'start': 124491972, 'end': 124491972, 'assembly': 'GRCh37', 'referenceBases': 'C', 'alternateBases': 'A', 'shareVariantLevelData': True}, 'zygosity': 1}]}}
2019-03-29 12:12:16 k1315C02VR0WRHTDH.local urllib3.connectionpool[11496] DEBUG Starting new HTTP connection (1): localhost:9020
INFO:patientMatcher.auth.auth:Authorized client with id "scout_test" submits a POST request
INFO:patientMatcher.parse.patient:Validating against SCHEMA #/definitions/request
INFO:patientMatcher.utils.phenotype:https://monarchinitiative.org/analyze/phenotypes.json?input_items=HP:0002664+HP:0000014+HP:0001222+HP:0009733+HP:0000510+HP:0001251+HP:0000118&limit=20&target_species=9606
INFO:patientMatcher.utils.phenotype:Sending HPO terms to Monarch service..
INFO:patientMatcher.utils.phenotype:server returned 100 phenotypes.
INFO:patientMatcher.parse.monarch:Monarch hits parsed successfully.
ERROR:patientMatcher.match.handler:Could't find any connected MME nodes. Aborting external matching.
ERROR:flask.app:Exception on /patient/add [POST]
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask_negotiate.py", line 14, in wrapper
    return fn(*args, **kwargs)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/flask_negotiate.py", line 27, in wrapper
    return fn(*args, **kwargs)
  File "/Users/dannil/sandbox/patientMatcher/patientMatcher/server/views.py", line 34, in add
    match_external=True, host=current_app.config.get('MME_HOST'))
  File "/Users/dannil/sandbox/patientMatcher/patientMatcher/utils/add.py", line 83, in backend_add_patient
    mongo_db['matches'].insert_one(matching_obj)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/pymongo/collection.py", line 676, in insert_one
    common.validate_is_document_type("document", document)
  File "/usr/local/miniconda3/envs/scout4_36/lib/python3.6/site-packages/pymongo/common.py", line 434, in validate_is_document_type
    "collections.MutableMapping" % (option,))
TypeError: document must be an instance of dict, bson.son.SON, bson.raw_bson.RawBSONDocument, or a type that inherits from collections.MutableMapping
INFO:werkzeug:127.0.0.1 - - [29/Mar/2019 12:12:18] "POST /patient/add HTTP/1.1" 500 -
2019-03-29 12:12:18 k1315C02VR0WRHTDH.local urllib3.connectionpool[11496] DEBUG http://localhost:9020 "POST /patient/add HTTP/1.1" 500 291
2019-03-29 12:12:18 k1315C02VR0WRHTDH.local scout.utils.matchmaker[11496] INFO An error occurred while sending HTTP request to server (Expecting value: line 1 column 1 (char 0))
2019-03-29 12:12:18 k1315C02VR0WRHTDH.local werkzeug[11496] INFO 127.0.0.1 - - [29/Mar/2019 12:12:18] "POST /cust000/643594/mme_add HTTP/1.1" 302 -
```
The case does get added to the pmatcher mongo, but scout registers the error and does not set it as Mme_submitted.

northwestwitch commented:
I'll try to re-create the error starting from an installation on the MME and Scout on a new environment

dnil commented:
Is it working for you with the current code? If so, it could also be some interesting difference in my setup..

northwestwitch commented:
What do you have in the collection named "nodes" in patientMatcher?

On Fri, Mar 29, 2019 at 3:22 PM Daniel Nilsson <notifications@github.com>
wrote:

> Is it working for you with the current code? If so, it could also be some
> interesting difference in my setup..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Clinical-Genomics/scout/pull/1152#issuecomment-478015677>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Aayssvb2uCMQEhCrwCmLHmOWT2VWvBgtks5vbiGrgaJpZM4a_Ah_>
> .
>

dnil commented:
I don’t have it at all: not created, since no nodes added..

northwestwitch commented:
Good, it should be so. I'll let you know, thanks!

On Fri, Mar 29, 2019 at 3:24 PM Daniel Nilsson <notifications@github.com>
wrote:

> I don’t have it at all: not created, since no nodes added..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Clinical-Genomics/scout/pull/1152#issuecomment-478016547>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Aaysso-IrLI9SImghCTIPJOviagisufeks5vbiI0gaJpZM4a_Ah_>
> .
>

dnil commented:
:+1: Thanks! I'll get out running now; will check in later.

northwestwitch commented:
Don't worry, we can take it next week as well. Enjoy the beautiful weather!
:)

On Fri, Mar 29, 2019 at 3:27 PM Daniel Nilsson <notifications@github.com>
wrote:

> 👍 Thanks! I'll get out running now; will check in later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Clinical-Genomics/scout/pull/1152#issuecomment-478017337>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AaysssvWwU0CfXZ3bHzECnYtVKlqrwZ4ks5vbiLDgaJpZM4a_Ah_>
> .
>

northwestwitch commented:
Ok I understood exactly what is wrong here. I've set up patientMatcher to automatically run matches on external nodes when you submit a new patient. But the poor things fails because of course there are no external nodes connected yet. Should have considered the option.. I'll fix it!

northwestwitch commented:
Ok I think I fixed the specific error. I've also added a couple of improvements to this branch on Scout.

To do:
1) Remove any patients from patients collection in pmatcher database.
2) Clone and reinstall PatientMatcher

dnil commented:
Oki; well done! Looks operational now; I can add cases w/o problem. Will read and test a bit more at a later point in time.. Cheers!

northwestwitch commented:
@dnil I've been thinking about the discussion of the other day: that users not allowed to submit to MME or whenever MME params are missing in the conf file, then nothing related to MME should be shown.

I make some changes and I think this is in place now. Places where MME related code could be shown:

- **Case view** (in this case MME form is shown only if user has 'mme_submitter' role, regardless of MME-params in the conf file. But if those params are missing then you get a flash error message which tells you to check the MME config params

- **matchmaker_matches** , **matchmaker_match**,  **matchmaker_add**,  **matchmaker_delete** views: Checks if user has 'mme_submitter' role and if MME-related params are in config file, otherwise redirects

dnil commented on scout/server/blueprints/cases/templates/cases/matchmaker.html:
```diff
@@ -0,0 +1,236 @@
+                        {{key}}:<strong>{{value.label}}</strong>
+                        <br>
+                      {% else %} <!-- this will be zygosity -->
+                        {{key}}:<strong>{{value}}</strong>
```
Is this way of describing zygosity a matchMaker standard? Otherwise I would suggest casting this as a string het / hom / hemi. I'm sure our users can learn, but this at least needs documentation. Also, flagging hemis (at least when homozygous state on X for a male) will be very useful, but could wait until a later release if it is complicated with the current setup!

northwestwitch replied:
The standard is defined in this document: https://github.com/ga4gh/mme-apis/blob/master/search-api.md

It says: 
zygosity: <number> (1 for heterozygous or hemizygous, 2 for homozygous)

You can't really distinguish between heteroz and hemiz because they both have 1. Didn't know how to render it from out data..

dnil replied:
Slightly unusual - we usually clump hom with hemi, but confirmed: as long as it is in their standard, lets keep it. If you are feeling *very* user friendly, add a string tag next to it (1 is "hemi" for variants on X for males, "het" otherwise, and 2 can always get a "hom"). 👍

northwestwitch replied:
Good idea, you are right, it's not obvious! Fixed!

dnil commented on scout/parse/matchmaker.py:
```diff
@@ -0,0 +1,226 @@
+            zygosities = var['samples'] # it's a list with zygosity situation for each sample of the case
+            for zyg in zygosities:
+                if zyg.get('display_name') == sample_name: # sample of interest
+                    zygosity = zyg['genotype_call'].count('1') + zyg['genotype_call'].count('2')
```
Thanks!

dnil replied:
The only thing that can go wrong noe is a het 1/2.. 🤣

northwestwitch replied:
Right? There won't be many of these cases but I guess they will be counted as homozygous. The user will need to check

dnil commented on scout/parse/matchmaker.py:
```diff
@@ -0,0 +1,226 @@
+            for zyg in zygosities:
+                if zyg.get('display_name') == sample_name: # sample of interest
+                    zygosity = zyg['genotype_call'].count('1') + zyg['genotype_call'].count('2')
+            g_feature['zygosity'] = zygosity
```
I had a few more thoughts on this: see below!

dnil commented on scout/server/blueprints/cases/controllers.py:
```diff
@@ -457,13 +462,13 @@ def gene_variants(store, variants_query, page=1, per_page=50):
     my_institutes = list(inst['_id'] for inst in user_institutes(store, current_user))
 
-    log.debug("Institutes allowed: {}.".format(my_institutes))
+    LOG.debug("Institutes allowed: {}.".format(my_institutes))
```
Do we actually need these LOG.debugs now? I think they are leftovers? Perhaps just drop them or make log entries at warning level?

northwestwitch replied:
I'll drop this and make a warning of those that actually are warnings

dnil replied:
Thanks - I'm not sure they were yours originally.

dnil commented on scout/server/blueprints/cases/controllers.py:
```diff
@@ -457,13 +462,13 @@ def gene_variants(store, variants_query, page=1, per_page=50):
         # hide other institutes for now
         if variant_obj['institute'] not in my_institutes:
-            log.debug("Institute {} not allowed.".format(variant_obj['institute']))
+            LOG.debug("Institute {} not allowed.".format(variant_obj['institute']))
```
This for instance may warrant a warning or info level entry?

dnil commented on scout/server/blueprints/cases/templates/cases/matchmaker.html:
```diff
@@ -0,0 +1,236 @@
+                                              {% if not ikey == 'shareVariantLevelData' %}
+                                                {{ikey}}:<strong>{{item}}</strong><br>
+                                              {% endif %}
+                                            {% endfor %}
```
This one lacks the distinction between 'type' and zygosity, doesn't it? I could not see zygosities of the matching patients. Perhaps either comment up this section: it is a little difficult to follow, or if possible make a subroutine that can be reused?

northwestwitch replied:
Right, I didn't realize. genomic features are objects like this:

![image](https://user-images.githubusercontent.com/28093618/55476209-434e8780-5616-11e9-952d-551223ff6e4e.png)

I'll fix it!

dnil commented on scout/server/blueprints/cases/views.py:
```diff
@@ -127,6 +127,187 @@ def generate_csv(header, lines):
+        flash('unauthorized request', 'warning')
+        return redirect(request.referrer)
+
+    # Required params for sending an add request to MME:
```
Is it worth it making a member method for the recurring things here as well? E.g. the population and checks are re-used almost the same in all of the methods.

northwestwitch replied:
You mean a method that collects these variables?
mme_base_url = current_app.config.get('MME_URL')
mme_accepts = current_app.config.get('MME_ACCEPTS')
mme_token = current_app.config.get('MME_TOKEN')

dnil replied:
Hm, yes, it looked like more. Still, the error checking simplifies into one call. One could make a single Mme object abstracting all of the options, but hey, we have to have something to refactor later.. 😉

northwestwitch replied:
I agree!

dnil commented on tests/conftest.py:
```diff
@@ -1224,3 +1225,145 @@ def hpo_genes(request, hpo_genes_handle):
     return parse_hpo_genes(hpo_genes_handle)
+
+
+#############################################################
```
Wow! Great!

dnil commented on tests/server/blueprints/cases/test_cases_views.py:
```diff
@@ -102,3 +103,104 @@ def test_causatives(app, user_obj, institute_obj):
+        )
+        assert store.case_collection.find({'mme_submission':{'$exists' : True}}).count() == 1
+
+        current_app.config['MME_URL'] = 'http://localhost:9020'
```
Soo, but only if there is no server alive? Good for a virtual machine I guess, but may not work on a dev station? ;-)

northwestwitch replied:
What makes you think that on a local machine it actually works? Because it doesn't

dnil replied:
Hm, I was thinking if the developer already has a service running on 9020? But perhaps this will fail already before that?

northwestwitch replied:
If you want to do the things properly then you should test using a server-service that dynamically gets a free port, such as I did when testing the controllers. For the purpose of this test (just showing that there is a views method with that name that you can call, you could write whatever string in that address and it won't affect the test). Which fake port do you like? :)

dnil replied:
Hm, tempting.. 😄 Perhaps just mention that in the test "given" statement: I don't think we need an issue for making a fake server-service etc.

northwestwitch replied:
I wrote it this way now
# Given mock MME connection parameters
current_app.config['MME_URL'] = 'http://fakey_mme_url:fakey_port' 

It felt like there are already controllers that are checking well all the processes related to MME so I didn't add the the fake service here as well, but we could always do it later maybe?

dnil commented on tests/server/blueprints/cases/test_matchmaker_controllers.py:
```diff
@@ -0,0 +1,187 @@
+MME_TOKEN = 'test_token'
+
+# mock matchmaker server to send responses when it get requests
+class MockServerRequestHandler(BaseHTTPRequestHandler):
```
Again, wow! That is beyond due diligence - really solid!

northwestwitch replied:
Right? I feel like I'm ready for the pension now :)

dnil replied:
👵 👍

dnil commented:
All ready to merge, aren't we? Documentation and changelog perhaps? What you had in some of the PR posts for installing patientMatcher + links and possibly info from the patientMatcher repo could go into INSTALLATION as an optional feature. On the USER side, particularly care for what gets submitted where and when: this is where we will get most worry and confusion - and joy, after the first few matches roll in. Also that link to the matchmaker API definition would be good to save!

northwestwitch commented:
> All ready to merge, aren't we? Documentation and changelog perhaps? What you had in some of the PR posts for installing patientMatcher + links and possibly info from the patientMatcher repo could go into INSTALLATION as an optional feature. On the USER side, particularly care for what gets submitted where and when: this is where we will get most worry and confusion - and joy, after the first few matches roll in. Also that link to the matchmaker API definition would be good to save!

Will do that. Thanks again a for the review and feedback, I know it was a lot of code and concepts and I really really appreciate that you took the time to do it! 🥇

northwestwitch commented:
Had some issues with doc figures, fixed now. I'm ready to merge I'd say!
